docs/design/2018-10-20-join-reorder-dp-v1.md

```diff
@@ -0,0 +1,100 @@
+
+## Proposal
+
+I’ve explored the join reorder algorithms of Orca, Calcite, Spark, MySQL and Postgres. Using DP when the number of nodes are small is an operable strategy.
+
+I implemented two DP algorithms. You can find them [here](https://github.com/winoros/DP-CCP/tree/master). DP-CPP is introduced in [this paper](https://dl.acm.org/citation.cfm?id=1164207).
+
+My experiment result shows that DP-CCP only has a better performance than DP-SUB when the input is list and circle. It cannot beat the DP-SUB when input is star and clique. This is different with the experiment result posted in the paper. The main reason is that maintaining histograms of join online is very costly. If we use 1μs to get the statistics of one join, the most costly part of star graph will be statistics.
+
+So I suggest using DP-SUB as the first version of join reorder since the performance of DP-CCP is not good enough.
+
+The core DP-SUB algorithm is like the following code, a complete implementation is [here](https://github.com/winoros/DP-CCP/blob/dp-sub/dpsub.cpp):
+
+```
+func dp(int nodeCnt) {
+	for i := 0; i < nodeCnt; i++ {
+		bestPlan[1 << i] = planByNodeID[i]
+		bestCost[1 << i] = 0
+	}
+	// Iterate the sub from small to big.
+	// This can make sure that node set S1 is visited earlier than S2 if S1 belongs to S2.
+	for state := 1; state < (1 << nodeCnt); state++ {
+		if bits.OnesCount(i) == 1 || (nodeCnt not connected) {
+			continue // It’s a single node or not connected, skip it;
+		}
+		// This loop can enumerate all its subset.
+		for sub := (state - 1) & state； sub > 0; sub = (sub - 1) & state {
+			// Inner join’s left and right have no difference, so we only consider it once.
+			if sub > remain {
+				continue
+			}
+			// That the dp result is nil means the subset is not connected.
+			if bestPlan[sub] == nil || bestPlan[state ^ sub] == nil {
+				continue
+			}
+			join := buildJoin(bestPlan[sub], bestPlan[state^sub], edges between them)
+			if bestCost[state] > join.Cost {
+				bestPlan[state] = join
+				bestCost[state] = join.Cost
+			}
+		}
+	}
+}
+```
+
+Here the set of node is shown as an integer. If the i-th bit of its binary representation is 1, the i-th node will be in this set.
+
+## Rationale
+
+### How other systems solve the same issue?
+
+DP-SUB is the most commonlly used solution in open source databases.
+
+### What is the disadvantage of this design?
+
+DP-SUB is hard to prune invalid branches. 
+
+And it’s a little hard to enlarge its search space to involve the outer join efficiently.
+
+In a word, DP-SUB cannot be extended easily. The next version of our join reorder will be very likely a new one instead of enhancing DP-SUB.
+
+## Compatibility
+
+There’s no compatibility issue.
+
+## Implementation
+
+First, extract the join nodes. If the number of nodes is less than x, we go into DP-SUB.
+
+For DP-SUB:
+
+1. Enlarge its edges by constant propagation.
```

Comment:
If we do not consider `OtherCondition`, then constant propagation should remove edges from the join graph? e.g, `t1.a = t2.a and t2.a = 1` is converted to `t1.a = 1 and t2.a = 1`.

Reply:
One side condition will be pushed down to join already.
We don't need to hold them when do join reorder.

Reply:
Yes, I mean `Enlarge its edges by constant propagation` should be `Remove its edges by constant propagation`?

Reply:
Oh i got it.
This propagation mentioned here is mainly to solve the following case:
```
t1.a=t2.a and t2.a=t3.a
```
There's actually one edge `t1.a=t3.a` and may be a edge involved in best choice.

And back to your case, that convention eats one edge and may cause the final plan worse than we expected. We need to come up with a way to handle it.